### PR TITLE
ClusterSharding should clean its internal cache if region/proxy died

### LINKED ISSUE #6398: ClusterShardRegion of 1.5 Alpha version unable to restart
Version of Akka.NET? 1.5 alpha-3
Which Akka.NET Modules? ClusterSharding

**Describe the bug**

When I use this F# script and reference to akka nuget package version 1.5 alpha-3, once I stop the regionshard, I would unable to restart it again.

https://github.com/Horusiath/Akkling/blob/master/examples/cluster-sharding.fsx

**To Reproduce**

After execute the script, execute:
```
system1.Stop fac1.ShardRegion
let fac1 = entityFactoryFor system1 "printer" <| props (actorOf2 behavior)
let entity1 = fac1.RefFor "shard-1" "entity-1"
entity1 <! "hello"
```
This time hello would not appear.

**Expected behavior**
new shard should be created again and works to print hello (1.4.49 works)

**Actual behavior**
The terminated shardregion still hangs there.

**Additional context**

The only workaround so far I have is to remove all specified shardregion in cluster by the following code:
```
let system1 = System.create "cluster-system" (configWithPort 5000)
let cs = ClusterSharding.Get system1

let cdrfi = cs.GetType().GetField("_regions", BindingFlags.Instance ||| BindingFlags.NonPublic)
let cdpfi = cs.GetType().GetField("_proxies", BindingFlags.Instance ||| BindingFlags.NonPublic)

let cdr : ConcurrentDictionary<string, IActorRef> = unbox (cdrfi.GetValue(cs))
let cdp : ConcurrentDictionary<string, IActorRef> = unbox (cdpfi.GetValue(cs))

match cdr.TryRemove "printer" with
| true, iaref -> printfn "%A" iaref
| _ -> ()
```

### PR DESCRIPTION
**Pull request by Arkatufus**
Fixes #6398

## Changes
* Shift region and proxy cache update responsibility from `ClusterSharding` to `ClusterShardingGuardian`
* Clean cache when region/proxy actor died

### CODE REVIEW

File: src/contrib/cluster/Akka.Cluster.Sharding/ClusterSharding.cs [hunk 1]
```diff
@@ -302,7 +302,8 @@ public ClusterSharding(ExtendedActorSystem system)
                 var guardianName = system.Settings.Config.GetString("akka.cluster.sharding.guardian-name");
                 var dispatcher = system.Settings.Config.GetString("akka.cluster.sharding.use-dispatcher");
                 if (string.IsNullOrEmpty(dispatcher)) dispatcher = Dispatchers.InternalDispatcherId;
-                return system.SystemActorOf(Props.Create(() => new ClusterShardingGuardian()).WithDispatcher(dispatcher), guardianName);
+                return system.SystemActorOf(Props.Create(() => new ClusterShardingGuardian(_regions, _proxies))
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ClusterSharding.cs:305]: LGTM

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests/ShardRegionSpec.cs [hunk 2]
```diff
@@ -118,6 +119,31 @@ private IActorRef StartProxy(ActorSystem sys)
             return ClusterSharding.Get(sys).StartProxy(shardTypeName, null, extractEntityId, extractShardId);
         }
 
+        [Fact(DisplayName = "ClusterSharding must clean up its internal regions cache when ShardRegion actor died")]
+        public void ClusterShardingDeadShardRegionTest()
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding.Tests/ShardRegionSpec.cs:123]: LGTM

**Comment by Arkatufus** [comment]
Bugged when there are both region and proxy with the same type name, let me fix that

**Review by Aaronontheweb — APPROVED** [review]
LGTM - pending test suite passing

**Pull request merged**